src/backend/storage/smgr/smgr.c

```diff
@@ -581,7 +581,7 @@ smgrextend(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum,
 		   char *buffer, bool skipFsync)
 {
 	if (file_extend_hook)
-		(*file_extend_hook)(reln, forknum, blocknum, buffer);
+		buffer = (*file_extend_hook)(reln, forknum, blocknum, buffer);
```

Comment:
When will the returned buffer be freed?

Reply:
This is a thread local, or long live in process memory.
```
// process local buffer
static char file_write_hook_buffer[BLCKSZ]                       = {'\0'};
static char ao_file_write_hook_buffer[MAX_APPENDONLY_BLOCK_SIZE] = {'\0'};
```

no need to free